File: src/main/java/hzt/model/entity/Path.java
package hzt.model.entity;

import javafx.collections.ObservableList;
import javafx.geometry.Point2D;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.paint.Paint;
import javafx.scene.shape.Line;

import static hzt.controller.AnimationService.LINE_STROKE_WIDTH;

public class Path extends Group {

    private Paint paint;

    public Path() {
        super();
    }

    public void addLine(Point2D currentPosition, Point2D prevPosition) {
        if (!currentPosition.equals(Point2D.ZERO) && !prevPosition.equals(Point2D.ZERO)) {
            Line line = new Line(prevPosition.getX(), prevPosition.getY(), currentPosition.getX(), currentPosition.getY());
            line.setStroke(paint);
            line.setStrokeWidth(LINE_STROKE_WIDTH);
            getChildren().add(line);
        }
    }

   public void fadeOut() {
        int size = getChildren().size();
        ObservableList<Node> children = getChildren();
        for (int i = 0, childrenSize = children.size(); i < childrenSize; i++) {
            Node node = children.get(i);
            Line line = (Line) node;
            line.setOpacity((float) i / size);
        }
    }

    public void removeLine(int index) {
        getChildren().remove(index);
    }

    public ObservableList<Node> getElements() {
        return getChildren();
    }

    public void setStroke(Paint paint) {
        this.paint = paint;
        getChildren().forEach(l -> ((Line) l).setStroke(paint));
    }

}

File: src/main/java/hzt/model/entity/Path.kt
package hzt.model.entity

import javafx.collections.ObservableList
import javafx.geometry.Point2D
import javafx.scene.Group
import javafx.scene.Node
import javafx.scene.paint.Paint
import javafx.scene.shape.Line
import java.util.function.Consumer

class Path : Group() {
    private var lineWidth = 0.0
    private var paint: Paint? = null
    fun addLine(currentPosition: Point2D, prevPosition: Point2D) {
        if (currentPosition != Point2D.ZERO && prevPosition != Point2D.ZERO) {
            val line = Line(prevPosition.x, prevPosition.y, currentPosition.x, currentPosition.y)
            line.stroke = paint
            line.strokeWidth = lineWidth
            children.add(line)
        }
    }

    fun setLineWidth(lineWidth: Double) {
        this.lineWidth = lineWidth
    }

    fun fadeOut() {
        val children = children
        val size = children.size
        for (i in 0 until size) {
            val node = children[i]
            val line = node as Line
            line.opacity = (i.toFloat() / size).toDouble()
        }
    }

    fun removeLine(index: Int) {
        children.removeAt(index)
    }

    val elements: ObservableList<Node>
        get() = children

    fun setStroke(paint: Paint?) {
        this.paint = paint
        children.forEach(Consumer { l: Node -> (l as Line).stroke = paint })
    }
}
